### PR TITLE
LLaMA in TF/Keras

### PR DESCRIPTION
**Pull request by t-kalinowski**

### CODE REVIEW

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 1]
```diff
@@ -0,0 +1,1198 @@
+---
+title: "LLaMA in R with Keras and TensorFlow"
+description: >
+  Walk-through an implementation of LLaMA in R
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:4]: Hesitating to make suggestions to a "much more native" speaker, but this sounds a bit strange to me... Maybe add "of", or make a complete sentence of it? "This post walks ..."

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 2]
```diff
@@ -0,0 +1,1198 @@
...
+  with TensorFlow and Keras.
+author:
+  - name: Tomasz Kalinowski
+    affiliation: RStudio
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:8]: guess this should be Posit :-)

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 3]
```diff
@@ -0,0 +1,1198 @@
...
+    toc: true
+categories:
+  - TensorFlow/Keras
+  - R
+  - Generative Models
+  - Natural Language Processing
+preview: images/preview.png
+---
+
+OpenAI’s chatGPT has awakened a collective awareness of what Large
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:24]: "awaken awareness" sounds strange to me, - "create"?

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 4]
```diff
@@ -0,0 +1,1198 @@
...
+Language Models (LLMs) are capable of. With that awakening comes a daily
+march of LLM news: new products, new features, new models, new
+capabilities, (and new worries). It seems we’re in the early parts of a
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:27]: leave out the comma?

Comment (skeydan) [reply _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:27]: (because parens)

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 5]
```diff
@@ -0,0 +1,1198 @@
...
+cambrian explosion of LLMs and LLM powered tools; it’s not yet clear how
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:28]:
```suggestion
Cambrian explosion of LLMs and LLM powered tools; it’s not yet clear how
```

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 6]
```diff
@@ -0,0 +1,1198 @@
...
+understanding first, capability second.
+
+Why LLaMA? With the sheer volume of LLM related content and news out
+there, it can seem daunting to know where to get started. Almost weekly
+it seems there is a new model announced. Browsing some hubs of LLM
+activity ([HuggingFace](https://huggingface.co/models),
+[TFHub](https://tfhub.dev/s?module-type=text-language-model),
+[reddit](https://www.reddit.com/r/deeplearning/),
+[HackerNews](https://hn.algolia.com/?q=LLM)) muddies the waters even
+more. How to pick?
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:52]: maybe add "a specific model"?

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 7]
```diff
@@ -0,0 +1,1198 @@
...
+Of the many LLM-related news items in the past months, one that stands
+head-and-shoulders above the crowd is the [release of
+LLaMA](https://ai.facebook.com/blog/large-language-model-llama-meta-ai/),
+a modern, foundational LLM made available to the public by Meta AI in
+Februay 2023. On common benchmarks, LLaMA outperforms OpenAI’s GPT-3,
+while being substantially smaller (though still *large*).
+
+LLaMA is a great starting place because it is a simple and modern
+architecture, has excellent performance on benchmarks, and is open. The
+model architecture has just a few new ideas incorporated into it since
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:63]: "has had"?

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 8]
```diff
@@ -0,0 +1,1198 @@
...
+way](https://forms.gle/jk851eBVbX1m5TAv5) or
+[another](https://github.com/facebookresearch/llama/pull/73). Note, the
+weights do come with their own license, which you can preview
+[here](https://github.com/facebookresearch/llama/pull/234).
+
+So, without further ado, let’s get started.
+
+### Setup
+
+First, we’ll want to install the required R and python packages, and
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:89]:
```suggestion
First, we’ll want to install the required R and Python packages, and
```

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 9]
```diff
@@ -0,0 +1,1198 @@
...
+  import_from(keras$layers, Dense)
+  np <- reticulate::import("numpy", convert = FALSE)
+
+  seq_len0 <- function(x) seq.int(from = 0L, length.out = x)
+})
+```
+
+If you’ve acquired the pre-trained weights, it’ll be convenient to
+convert them from the torch checkpoint format to something that’s more
+framework agnostic (you only need to do this once of course):
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:126]: comma? (", of course")

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 10]
```diff
@@ -0,0 +1,1198 @@
...
+
+show_tokens(prompt)
+```
+
+            1       450      1900       982       304     13978       367       267
+           ""     "The"    "best"     "way"      "to" "attract"      "be"      "es"
+
+Note that “bees” is two tokens. Not every token corresponds to a word.
+For example, one non-word token we can reliably expect to show up in a
+tokenizer trained on a corpus of english text is “ing”. However when the
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:234]:
```suggestion
tokenizer trained on a corpus of English text is “ing”. However when the
```

Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:234]:
```suggestion
tokenizer trained on a corpus of english text is “ing”. However, _when_ the
```

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 11]
```diff
@@ -0,0 +1,1198 @@
...
+
+``` r
+show_tokens("wonking")
+```
+
+         1   2113   9292
+        ""  "won" "king"
+
+Another thing to note about the tokenizer is that each token sequence
+starts with token id `1`. This is a special *beginning-of-sequence*
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:268]:
```suggestion
starts with token id `1`. This is a special *begin-of-sequence*
```

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 12]
```diff
@@ -0,0 +1,1198 @@
...
+}
+```
+
+While there is not a lot of code, there are a lot of ideas packed in
+there. This block forms the main trunk of the model, so it’s worth
+taking the time to go through it slowly.
+
+We implement the `TransformerBlock` as a subclassed
+`keras.layers.Layer`. This is gives us some niceties like the ability to
+compose with other keras layers, but these are mostly orthogonal for the
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:453]:
```suggestion
compose with other keras layers, but these are mostly irrelevant to the
```

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 13]
```diff
@@ -0,0 +1,1198 @@
...
+``` r
+x2 <- x |> ...
+x <- x + x2 # add residual x to x2
+```
+
+This is a common pattern that helps with model training, and especially
+to help with the [vanishing gradient
+problem](https://en.wikipedia.org/wiki/Vanishing_gradient_problem). It’s
+a skip-connection in the other-wise linear sequence of matrix
+transformations. It reinjects information (during the forward pass), and
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:476]: That's a very nice explanation!

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 14]
```diff
@@ -0,0 +1,1198 @@
...
+In a moment we’ll see that that `feed_foward` is a slightly fancier
+variation of a conventional sequence of `Dense` layer. Before we get
+there we can we safely skip ahead to distill the following intuition: a
+`TransformerBlock` is basically an `Attention` layer followed by a few
+(fancy) dense layers, with some simple composition patterns (tricks)
+that help with training. `Attention` is the heart of the model: it’s the
+most interesting, and also the most involved.
+
+With the framing in place, let’s go through and take a closer look at
+`RMSNorm`, `FeedForward` and then with the foundation in place, we’ll
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:516]:
```suggestion
`RMSNorm`, `FeedForward`, and then with the foundation in place, we’ll
```

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 15]
```diff
@@ -0,0 +1,1198 @@
...
+    output <- self$wo(output) # (bsz, seqlen, n_heads * head_size)
+
+    output
+  }
+}
+```
+
+`Attention` in LLaMA is similar but not identical to the Attention
+described in the [original Transformers
+paper](https://arxiv.org/abs/1706.03762), (and available as a keras
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:779]: remove comma?

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 16]
```diff
@@ -0,0 +1,1198 @@
...
+
+  # broadcast over batch and heads dim
+  mask[tf$newaxis, tf$newaxis, , ] # (1, 1, seqlen, seqlen)
+}
+```
+
+The mask is a strictly upper triangular matrix filled with `-Inf`
+values. Adding the mask to the attention scores prevents the model from
+being able to “look ahead” and see the attention score for a token
+pairings it hasn’t seen yet at a particular position in the sequence.
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:860]: pairings doesn't belong there right?

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 17]
```diff
@@ -0,0 +1,1198 @@
...
+
+Next lets turn our attention to `apply_rotary_embedding()`. This core
+innovation was published by @su2022roformer in the paper titled
+["RoFormer: Enhanced Transformer with Rotary Position Embedding"](https://arxiv.org/abs/2104.09864).
+
+Some context:
+
+-   The bare `Attention()` mechanism doesn’t leave any possibility for a
+    token’s position in a sequence to affect the attention scores, since
+    only token-pairs are scored. Attention treats it’s input like a
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:883]:
```suggestion
    only token-pairs are scored. Attention treats its input like a
```

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 18]
```diff
@@ -0,0 +1,1198 @@
...
+> Specifically, incorporating the relative position embedding is
+> straightforward: simply rotate the affine-transformed word embedding
+> vector by amount of angle multiples of its position index and thus
+> interprets the intuition behind *Rotary Position Embedding*
+
+Expanding slightly: the rotation matrix is designed so that
+subsequently, after rotating our `q` and `k` token sequence embedding
+the same way, the *angle* between token features is a function of the
+relative distance between those tokens in the token sequence. The
+relative angle between two tokens is invariant to the absolution
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:906]:
```suggestion
relative angle between two tokens is invariant to the absolute
```

File: _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd [hunk 19]
```diff
@@ -0,0 +1,1198 @@
...
+
+### Wrapping up
+
+In this blog post we’ve walked through the LLaMA architecture
+implemented in R TensorFlow, including how to load pretrained weights,
+and then run the model to generate a sentence. Note, much of the code in
+this blog post is tailored for didactic purposes. While the
+implementation of the LLaMA architecture covered in this blog post is
+appropriate for training, there are a few modifications you’ll want to
+make before doing much text generation. Those include things like:
```
Comment (skeydan) [inline _posts/2023-05-24-llama-tensorflow-keras/llama-tensorflow-keras.Rmd:1169]:
```suggestion
make before doing a lot of text generation. Those include things like:
```

**Review by skeydan — APPROVED** [review]
Lookd great @t-kalinowski ! Very readable, and of interest to a wide range of users!

**Review by dfalbel — APPROVED** [review]
Looks great @t-kalinowski !

**Comment by t-kalinowski** [comment]
Thank you @skeydan and @dfalbel for the review and feedback!
